chemprop/uncertainty/uncertainty_calibrator.py

```diff
@@ -733,6 +730,281 @@ def nll(self, preds: List[List[float]], unc: List[List[float]], targets: List[Li
+
+    def nll(
+        self,
+        preds: List[List[float]],
+        unc: List[List[float]],
+        targets: List[List[float]],
+        mask: List[List[bool]],
+    ):
+        raise NotImplementedError
+
+    @staticmethod
+    def get_preds(predictor: UncertaintyPredictor):
+        return np.array(predictor.get_uncal_preds())
+
+    def calibrate(self):
+        uncal_preds = self.get_preds(self.calibration_predictor)  # shape(data, 2 * tasks)
+        targets = np.array(self.calibration_data.targets(), dtype=float)  # shape(data, tasks)
+        targets = np.nan_to_num(targets)
+
+        self.num_tasks = uncal_preds.shape[1] // 2
+        self.qhats = []
+
+        for task_id in range(self.num_tasks):
+            targets_task_id = targets[:, task_id]
+            uncal_preds_lower = uncal_preds[:, task_id]
+            uncal_preds_upper = uncal_preds[:, task_id + self.num_tasks]
+
+            calibration_scores = np.maximum(
+                uncal_preds_lower - targets_task_id, targets_task_id - uncal_preds_upper
+            )
+            calibration_scores = np.append(calibration_scores, np.inf)
+            calibration_scores = np.sort(np.absolute(calibration_scores))
+            self.qhats.append(np.quantile(calibration_scores, 1 - self.conformal_alpha / self.num_tasks))
+
+    def apply_calibration(self, uncal_predictor: UncertaintyPredictor):
+        uncal_preds = self.get_preds(uncal_predictor)  # shape(data, task)
+        num_data = uncal_preds.shape[0]
+        intervals = np.zeros((2 * self.num_tasks, num_data), dtype=float)
+
+        for task_id in range(self.num_tasks):
+            uncal_preds_lower = uncal_preds[:, task_id]
+            uncal_preds_upper = uncal_preds[:, task_id + self.num_tasks]
+
+            intervals[task_id] = uncal_preds_lower - self.qhats[task_id]
+            intervals[task_id + self.num_tasks] = uncal_preds_upper + self.qhats[task_id]
+
+        intervals = np.transpose(intervals)
+        return uncal_preds.tolist(), intervals.tolist()
+
+
+class ConformalRegressionCalibrator(ConformalQuantileRegressionCalibrator):
+    """
+    Conformal Calibrator for regression datasets. Outputs interval of fixed size, centered around
+    model prediction, for each datapoint. Intervals should cover 1-alpha proportion of datapoints.
+    As discussed in https://arxiv.org/abs/2107.07511.
+    """
+
+    @property
+    def label(self):
+        return f"{self.uncertainty_method}_conformal_regression"
+
+    def raise_argument_errors(self):
+        super().raise_argument_errors()
+        if self.dataset_type != "regression":
+            raise ValueError(
+                "Conformal Regression is only implemented for regression dataset types."
+            )
+
+    @staticmethod
+    def get_preds(predictor: UncertaintyPredictor):
+        preds = np.array(predictor.get_uncal_preds())
+
+        preds_duplicated = np.concatenate((preds, preds), axis=1)
+        return preds_duplicated
+
+
```

Comment:
There is no calibration going on here. This is not a calibrator.

If I'm understanding correctly, this is intended to be used with the quantile_interval loss function and report the bounds from that as the bounds here.

I would suggest instead to make this an uncertainty_method instead of a calibrator. Because it's a symmetrical interval, return the center of the interval as the predicted value. Return the 1/2 range of the interval as the interval value in the uncertainty.

Reply:
I think we can keep the output as upper and lower bounds for now, and we will probably switch to return the center of the interval as the predicted value in v2.

Reply:
That's going to make the output inconsistent with other uncertainty methods. It will be its own special thing and we'd need to make sure it's not used for calibration or evaluations.

Reply:
@cjmcgill I just adjusted the code to make it consistent with other uncertainty methods.